File: backend-app/src/routes/agentTools.ts
import { Elysia } from "elysia";
import { env } from "../config/env.js";
import {
  listBookedTimes,
  setCallLog,
  setCallbackAfter,
  setIntakeNotes,
  updateIntakeStatus,
} from "../db/intakes.js";
import { formatSpoken, joinSpoken } from "../lib/spoken.js";
import {
  checkAvailability,
  localDateOf,
  normalizeDateTime,
  slotsForDate,
  suggestSlots,
} from "../schedule/slots.js";
import { bookExisting, createAndBook } from "../services/booking.js";

// Agent tools — the "custom functions" the voice agent (Retell) calls DURING a call to
// check availability, list openings, and book. Retell POSTs a JSON body; in "standard"
// mode it's `{ name, call, args }`, in "args only" mode the arguments are at the top
// level. `readArgs` handles both. Every response is a JSON object with a speakable
// `message` plus structured fields, and always returns 200 so the agent can recover
// conversationally (auth failures are the only non-2xx). See docs/agent-tools.md.

interface RetellCall {
  from_number?: string;
  retell_llm_dynamic_variables?: Record<string, unknown>;
}

// Pull the function arguments whether Retell nested them under `args` or flattened them.
function readArgs(body: unknown): Record<string, unknown> {
  if (body && typeof body === "object") {
    const b = body as Record<string, unknown>;
    if (b.args && typeof b.args === "object") return b.args as Record<string, unknown>;
    return b;
  }
  return {};
}

function readCall(body: unknown): RetellCall {
  if (body && typeof body === "object") {
    const c = (body as Record<string, unknown>).call;
    if (c && typeof c === "object") return c as RetellCall;
  }
  return {};
}

const str = (v: unknown): string => (typeof v === "string" ? v.trim() : "");

// Parse a finite number from an unknown value (Retell may send it as a number or a string).
const num = (v: unknown): number | null => {
  if (typeof v === "number" && Number.isFinite(v)) return v;
  if (typeof v === "string" && v.trim() !== "" && Number.isFinite(Number(v))) return Number(v);
  return null;
};

const TZ = env.schedule.timezone;

export const agentTools = new Elysia({ prefix: "/agent" })
  // Shared-secret guard. When AGENT_TOOLS_SECRET is set, require a matching header
  // (configure it as a custom header on the Retell function). Unset = open (dev).
  .onBeforeHandle(({ headers, status }) => {
    if (env.agentToolsSecret && headers["x-agent-secret"] !== env.agentToolsSecret) {
      return status(401, { error: "unauthorized" });
    }
  })

  // Is a specific time open? If not, offer the nearest alternatives. arg: dateTime (ISO).
  .post("/check-availability", async ({ body }) => {
    const args = readArgs(body);
    const call = readCall(body);
    // Prefer the LLM-provided arg, but fall back to the dateTime dynamic variable passed to
    // the call — the same reliability fix as /agent/book. (The real fix is binding the Retell
    // function's dateTime parameter to {{dateTime}} so the model doesn't guess a value.)
    const dyn = (call.retell_llm_dynamic_variables ?? {}) as Record<string, unknown>;
    const raw = str(args.dateTime) || str(dyn.dateTime);
    if (!raw || Number.isNaN(new Date(raw).getTime())) {
      return { available: false, message: "I couldn't read that date and time." };
    }
    // The agent sends a zone-less local time (Pacific); read it as Pacific, not UTC.
    const dateTime = normalizeDateTime(raw, TZ);
    const now = Date.now();
    const booked = await listBookedTimes();
    const { available, reason } = checkAvailability(env.schedule, dateTime, booked, now);
    const when = formatSpoken(dateTime, TZ);
    if (available) {
      return { available: true, when, message: `Good news — ${when} is available.` };
    }
    const suggestions = suggestSlots(env.schedule, dateTime, booked, {
      limit: 3,
      withinDays: 14,
      now,
    });
    const alternatives = suggestions.map((s) => formatSpoken(s.start, TZ));
    // A ready-to-speak, joined version of the alternatives ("A, B, or C") — a single string
    // the Retell flow can extract into a {{alternatives}} variable and read aloud. The raw
    // `alternatives` array is kept too (for any structured use).
    const alternativesText = joinSpoken(alternatives);
    const lead = reason === "in_past" ? `${when} is in the past.` : `${when} isn't available.`;
    return {
      available: false,
      when,
      reason,
      alternatives,
      alternativesText,
      message: alternatives.length
        ? `${lead} The closest openings I have are ${alternativesText}.`
        : `${lead} I don't have anything open nearby — would another day work?`,
    };
  })

  // A few open slots on a given day. arg: date (YYYY-MM-DD).
  .post("/openings", async ({ body }) => {
    const args = readArgs(body);
    const call = readCall(body);
    const dyn = (call.retell_llm_dynamic_variables ?? {}) as Record<string, unknown>;
    // Anchor = the day (and, if given, the time) the caller cares about. One arg handles both
    // "Monday" and "Monday around 3 PM": we list that day's openings ordered by closeness to the
    // anchor, so a bare day surfaces the earliest slots while "around 3 PM" surfaces mid-afternoon
    // ones. Accept dateTime (preferred — carries a time) or a bare date; fall back to the
    // originally requested time.
    const anchorRaw = str(args.dateTime) || str(args.date) || str(dyn.dateTime);
    const anchorIso = anchorRaw ? normalizeDateTime(anchorRaw, TZ) : "";
    const anchor = anchorIso ? new Date(anchorIso) : null;
    if (!anchor || Number.isNaN(anchor.getTime())) {
      return { openings: [], message: "Which day would you like me to check?" };
    }
    const date = localDateOf(TZ, anchor.getTime());
    const anchorMs = anchor.getTime();

    const now = Date.now();
    const booked = await listBookedTimes();
    const day = slotsForDate(env.schedule, date, booked);
    const nearest = day.slots
      .filter((s) => s.available && new Date(s.start).getTime() > now)
      // Pick the 5 slots closest to the anchor time…
      .sort(
        (a, b) =>
          Math.abs(new Date(a.start).getTime() - anchorMs) -
          Math.abs(new Date(b.start).getTime() - anchorMs),
      )
      .slice(0, 5)
      // …then read them back in chronological order.
      .sort((a, b) => new Date(a.start).getTime() - new Date(b.start).getTime());
    const open = nearest.map((s) => formatSpoken(s.start, TZ));
    // Joined, ready-to-speak string for the flow to extract into {{openings}} and read aloud.
    const openingsText = joinSpoken(open);
    return {
      date,
      openings: open,
      openingsText,
      message: open.length
        ? `I have ${openingsText}.`
        : "I don't have any openings that day. Would another day work?",
    };
  })

  // Book a slot. For an inbound caller (no intakeId) we create the lead first. Booking
  // also creates the Cal.com meeting (invite + link). args: dateTime (ISO) + either
  // intakeId (existing lead) OR name/email[/phone/language] (new inbound lead).
  .post("/book", async ({ body }) => {
    const args = readArgs(body);
    const call = readCall(body);
    // Values the LLM extracts arrive in `args`; values passed to the call as dynamic
    // variables (intake_id, lead_name, email…) live on `call.retell_llm_dynamic_variables`.
    // Prefer args, but fall back to the dynamic variables — the LLM often does NOT re-emit a
    // hidden variable like intake_id as a function parameter, so reading it from the call is
    // the reliable path. (Requires standard payload mode so the `call` object is sent.)
    const dyn = (call.retell_llm_dynamic_variables ?? {}) as Record<string, unknown>;
    const raw = str(args.dateTime) || str(dyn.dateTime);
    if (!raw || Number.isNaN(new Date(raw).getTime())) {
      return { booked: false, message: "I couldn't read that date and time." };
    }
    // The agent sends a zone-less local time (Pacific); read it as Pacific, not UTC.
    const dateTime = normalizeDateTime(raw, TZ);
    const when = formatSpoken(dateTime, TZ);
    const intakeId =
      str(args.intakeId) || str(args.intake_id) || str(dyn.intake_id) || str(dyn.intakeId);

    const result = intakeId
      ? await bookExisting(intakeId, dateTime)
      : await bookNewLead(args, dyn, call, dateTime);

    if (!result.ok) {
      if (result.reason === "not_found") {
        return { booked: false, reason: result.reason, message: "I couldn't find that booking." };
      }
      if (result.reason === "in_past") {
        return { booked: false, reason: result.reason, message: `${when} is in the past.` };
      }
      if (result.reason === "missing_details") {
        return { booked: false, reason: result.reason, message: "I still need your name and email to book this." };
      }
      return {
        booked: false,
        reason: result.reason,
        message: `${when} was just taken. Would another time work?`,
      };
    }
    const email = result.intake.email;
    return {
      booked: true,
      when,
      intakeId: result.intake.id,
      message: `You're all set for ${when}. A confirmation with the meeting link will be sent to ${email}.`,
    };
  })

  // Schedule a callback: the person answered but wants to be reached later. We already have
  // their number (we dialed it), so we just persist WHEN to try again — the poller holds the
  // lead until then instead of using its default delay. args: callbackAfter (ISO local time)
  // + intakeId (which lead). Like the others, intakeId falls back to the call's dynamic vars.
  .post("/callback", async ({ body }) => {
    const args = readArgs(body);
    const call = readCall(body);
    const dyn = (call.retell_llm_dynamic_variables ?? {}) as Record<string, unknown>;
    const intakeId =
      str(args.intakeId) || str(args.intake_id) || str(dyn.intake_id) || str(dyn.intakeId);
    if (!intakeId) {
      return { scheduled: false, message: "I couldn't find your record to schedule a callback." };
    }
    // Relative callback ("in 10 minutes", "in an hour"): the agent extracts just the number of
    // minutes and we compute from server-now — accurate at the moment of the call, since the
    // agent can't reliably know the current clock time. Absolute times ("tomorrow at 2") still
    // come in as `callbackAfter`, read as Pacific wall-clock.
    const minutes =
      num(args.callbackInMinutes) ?? num(args.callback_in_minutes) ?? num(dyn.callbackInMinutes);
    let callbackAfter: string;
    if (minutes !== null && minutes > 0) {
      callbackAfter = new Date(Date.now() + minutes * 60_000).toISOString();
    } else {
      const raw =
        str(args.callbackAfter) || str(args.callback_after) || str(args.dateTime) || str(dyn.callbackAfter);
      if (!raw || Number.isNaN(new Date(raw).getTime())) {
        return { scheduled: false, message: "When would be a good time for us to call you back?" };
      }
      callbackAfter = normalizeDateTime(raw, TZ);
    }
    const when = formatSpoken(callbackAfter, TZ);
    const record = await setCallbackAfter(intakeId, callbackAfter);
    if (!record) {
      return { scheduled: false, message: "I couldn't find your record to schedule a callback." };
    }
    return {
      scheduled: true,
      when,
      message: `Got it — we'll reach back out around ${when}.`,
    };
  })

  // Stop calling a lead when a call hits a dead end a retry won't fix — a WRONG NUMBER (the
  // number doesn't reach the lead → unreachable) or a DECLINE (we reached them and they said no
  // → contacted). Moving the lead out of "new" is what makes the poller stop re-dialing. The
  // agent passes `outcome` as a fixed const per flow path, so it's deterministic. Keyed to the
  // lead by intakeId (args or dynamic vars, same as the other tools).
  .post("/mark-outcome", async ({ body }) => {
    const args = readArgs(body);
    const call = readCall(body);
    const dyn = (call.retell_llm_dynamic_variables ?? {}) as Record<string, unknown>;
    const intakeId =
      str(args.intakeId) || str(args.intake_id) || str(dyn.intake_id) || str(dyn.intakeId);
    if (!intakeId) {
      return { ok: false, message: "I don't have a record to update." };
    }
    const outcome = (str(args.outcome) || "unreachable").toLowerCase();
    const OUTCOMES: Record<string, { status: "unreachable" | "contacted"; note: string }> = {
      wrong_number: { status: "unreachable", note: "Wrong number — does not reach the lead." },
      unreachable: { status: "unreachable", note: "Marked unreachable." },
      declined: { status: "contacted", note: "Declined the consultation." },
    };
    const chosen = OUTCOMES[outcome] ?? {
      status: "unreachable" as const,
      note: "Marked unreachable.",
    };
    const record = await updateIntakeStatus(intakeId, chosen.status);
    if (!record) {
      return { ok: false, message: "I couldn't find that record." };
    }
    await setIntakeNotes(intakeId, chosen.note);
    return { ok: true, message: "Noted — we won't keep calling this lead." };
  })

  // Save a finished call's transcript (+ optional one-line summary) so it's reviewable per lead
  // in the dashboard. Called once by the realtime agent when the call ends. Keyed by intakeId.
  .post("/call-log", async ({ body }) => {
    const args = readArgs(body);
    const call = readCall(body);
    const dyn = (call.retell_llm_dynamic_variables ?? {}) as Record<string, unknown>;
    const intakeId =
      str(args.intakeId) || str(args.intake_id) || str(dyn.intake_id) || str(dyn.intakeId);
    if (!intakeId) {
      return { ok: false, message: "No record to attach the transcript to." };
    }
    const record = await setCallLog(intakeId, str(args.transcript), str(args.summary) || null);
    if (!record) {
      return { ok: false, message: "I couldn't find that record." };
    }
    return { ok: true };
  });

// Create a new inbound lead from the call details, then book it. Reads from the LLM args
// first, then the call's dynamic variables (same reliability reasoning as intake_id above).
async function bookNewLead(
  args: Record<string, unknown>,
  dyn: Record<string, unknown>,
  call: RetellCall,
  dateTime: string,
) {
  const name = str(args.name) || str(dyn.lead_name) || str(dyn.name);
  const email = str(args.email) || str(dyn.email);
  const phone = str(args.phone) || str(dyn.phone) || str(call.from_number);
  if (!name || !email) return { ok: false as const, reason: "missing_details" };
  return createAndBook(
    {
      language: str(args.language) || str(dyn.language) || "English",
      name,
      email,
      phoneNumber: phone,
      purpose: str(args.purpose) || str(dyn.purpose) || "Consultation (inbound call)",
      // The requested day is the date of the slot they're booking (in the business timezone).
      requestedDate: localDateOf(TZ, new Date(dateTime).getTime()),
    },
    dateTime,
  );
}

File: openai-agent-app/src/openai_agent/tools/agent_tools.py
"""The agent's tools — the same shared backend `/agent/*` endpoints the Retell agent calls.

`TOOL_SCHEMAS` are the function definitions the Realtime model sees. `ToolExecutor` runs a
requested call against the backend, injecting the lead's `intake_id` from the call context
(never from the model, so it can't be guessed wrong — the same reliability lesson as the Retell
`const {{intake_id}}` binding). Each handler returns a JSON string the model reads back.
"""

from __future__ import annotations

import json
import logging

import httpx

from ..config import Config

log = logging.getLogger(__name__)

# Function definitions for the Realtime session. Mirror the Retell functions; the model supplies
# the spoken details (a time, a day, an outcome), and the executor adds intake_id + the header.
TOOL_SCHEMAS: list[dict] = [
    {
        "type": "function",
        "name": "check_availability",
        "description": "Check whether a specific date and time is open for a 30-minute "
        "consultation. Call before confirming a time with the lead.",
        "parameters": {
            "type": "object",
            "properties": {
                "dateTime": {
                    "type": "string",
                    "description": "The time to check, ISO 8601. Resolve the spoken time against "
                    "today's date (Pacific); for a relative day like 'tomorrow' or 'next Monday', "
                    "use that day's actual future date, NOT today's.",
                }
            },
            "required": ["dateTime"],
        },
    },
    {
        "type": "function",
        "name": "get_openings",
        "description": "List a few open times on a day, ordered around the time the lead cares "
        "about. Use when they want to see options or gave an approximate time.",
        "parameters": {
            "type": "object",
            "properties": {
                "dateTime": {
                    "type": "string",
                    "description": "The day to list, ISO 8601, INCLUDING the approximate time the "
                    "lead mentioned so the openings come back around it — e.g. 'around 3 PM "
                    "tomorrow' -> that day at 15:00. If they named only a day with no time, use "
                    "12:00 (noon) of that day. Resolve relative days ('tomorrow', 'next Monday') "
                    "to their actual future date (Pacific), not today.",
                }
            },
            "required": ["dateTime"],
        },
    },
    {
        "type": "function",
        "name": "book_appointment",
        "description": "Book the consultation once the lead has agreed to a specific time.",
        "parameters": {
            "type": "object",
            "properties": {
                "dateTime": {"type": "string", "description": "The agreed slot, ISO 8601."}
            },
            "required": ["dateTime"],
        },
    },
    {
        "type": "function",
        "name": "schedule_callback",
        "description": "Record when to call the lead back (they answered but can't talk now). "
        "Use callback_in_minutes for relative times ('in 10 minutes'), else callback_after.",
        "parameters": {
            "type": "object",
            "properties": {
                "callback_in_minutes": {
                    "type": "number",
                    "description": "Minutes from now to call back, for relative requests.",
                },
                "callback_after": {
                    "type": "string",
                    "description": "Absolute time to call back, ISO 8601.",
                },
            },
        },
    },
    {
        "type": "function",
        "name": "mark_outcome",
        "description": "Record a dead-end outcome so we stop calling this lead: a wrong number "
        "or a decline.",
        "parameters": {
            "type": "object",
            "properties": {
                "outcome": {
                    "type": "string",
                    "enum": ["wrong_number", "declined", "unreachable"],
                    "description": "What happened.",
                }
            },
            "required": ["outcome"],
        },
    },
    {
        "type": "function",
        "name": "end_call",
        "description": "Hang up the phone. Call this ONLY after you've said your final words — a "
        "farewell once the lead is done, or your brief sign-off for a wrong number, voicemail, or "
        "decline. Saying goodbye does not hang up by itself; this is what actually ends the call.",
        "parameters": {"type": "object", "properties": {}},
    },
]


# The INBOUND screening agent gets a different tool set, not a superset. A receptionist answering a
# stranger has no lead to book against, no callback to schedule and no lead status to mark — it
# hands off, takes a message, or hangs up. Sharing one list would offer the model five tools it can
# never legitimately use on an inbound call, which is how agents end up booking phantom leads.
INBOUND_TOOL_SCHEMAS: list[dict] = [
    {
        "type": "function",
        "name": "transfer_to_human",
        "description": "Hand this caller to a real person. Use ONLY for a genuine request to "
        "book, schedule, or meet with someone, or when the caller directly asks for a human. "
        "Never for a general question, a complaint, or a sales call.",
        "parameters": {
            "type": "object",
            "properties": {
                "reason": {
                    "type": "string",
                    "description": "One sentence, in ENGLISH, describing what the caller wants. "
                    "It is read aloud to the colleague before they accept the call, so make it "
                    "specific — e.g. 'Wants to book a consultation about a logistics AI project.'",
                }
            },
            "required": ["reason"],
        },
    },
    {
        "type": "function",
        "name": "take_message",
        "description": "Record a message for the team when the caller wants a callback, or when "
        "you could not help them and no transfer is appropriate.",
        "parameters": {
            "type": "object",
            "properties": {
                "caller_name": {"type": "string", "description": "The caller's name, as given."},
                "callback_number": {
                    "type": "string",
                    "description": "The number to call back, digits only. Read it back to the "
                    "caller to confirm before recording it.",
                },
                "message": {
                    "type": "string",
                    "description": "What the call is regarding, in one or two sentences.",
                },
            },
            "required": ["message"],
        },
    },
    # Same contract as the outbound agent: the bridge speaks the farewell and closes the line.
    {
        "type": "function",
        "name": "end_call",
        "description": "Hang up the phone. Call this once the conversation is genuinely over — "
        "the caller is done, it was a wrong number, or it was a sales call you declined. A warm "
        "farewell is spoken automatically, so do not say goodbye yourself first.",
        "parameters": {"type": "object", "properties": {}},
    },
]


class ToolExecutor:
    """Runs a tool call for one lead against the shared backend."""

    def __init__(self, cfg: Config, *, intake_id: str) -> None:
        self._cfg = cfg
        self._intake_id = intake_id

    async def run(self, name: str, args: dict) -> str:
        """Execute the named tool and return a JSON string for the model to read back."""
        try:
            if name == "check_availability":
                return await self._post("/agent/check-availability", {"dateTime": args.get("dateTime", "")})
            if name == "get_openings":
                return await self._post("/agent/openings", {"dateTime": args.get("dateTime", "")})
            if name == "book_appointment":
                return await self._post(
                    "/agent/book", {"dateTime": args.get("dateTime", ""), "intakeId": self._intake_id}
                )
            if name == "schedule_callback":
                body: dict = {"intakeId": self._intake_id}
                if args.get("callback_in_minutes") is not None:
                    body["callbackInMinutes"] = args["callback_in_minutes"]
                if args.get("callback_after"):
                    body["callbackAfter"] = args["callback_after"]
                return await self._post("/agent/callback", body)
            if name == "mark_outcome":
                return await self._post(
                    "/agent/mark-outcome",
                    {"intakeId": self._intake_id, "outcome": args.get("outcome", "unreachable")},
                )
            # Retained so a re-added schema still works, but note_caller is no longer offered:
            # every call cost the caller a turn. The model would speak, call the tool, then start a
            # NEW response after the result — "Hi Michael, let me get oriented" followed by "Hi
            # Michael. What can I help you with?". The name is now read off the transcript instead,
            # which costs nothing and produces the sentence the caller should have heard first.
            if name == "note_caller":
                # Purely local, like take_message: nothing to post anywhere, it just needs to reach
                # the bridge's state so the finished call carries a name. Kept OUT of take_message
                # because a caller giving their name is not the same event as leaving a message —
                # most callers do the first and never the second.
                return json.dumps({"recorded": True})
            if name == "take_message":
                # Inbound only. There is no intake to attach this to (a stranger called us), so the
                # message rides out on the call log the bridge already persists at call end — see
                # `messages` in the bridge state. Returning it lets the model read the confirmation
                # back rather than inventing one.
                return json.dumps(
                    {
                        "recorded": True,
                        "message": "Message recorded — the team will follow up.",
                    }
                )
            if name in ("end_call", "transfer_to_human"):
                # Both are handled by the bridge, which owns the call's plumbing: end_call drains
                # the audio and closes the stream; transfer_to_human redirects the live call out of
                # the stream entirely (a `<Connect>` cannot be escaped from in-band). This branch is
                # only a safety net so neither tool ever looks "unknown" to the model.
                return json.dumps({"ok": True})
            return json.dumps({"error": f"unknown tool {name}"})
        except Exception as exc:  # noqa: BLE001 — a tool failure must not drop the call
            log.warning("tool %s failed: %s", name, exc)
            return json.dumps({"error": "That didn't go through — let's try again."})

    async def record_call_log(self, transcript: str, summary: str) -> None:
        """Persist the finished call's transcript (+ one-line summary) to the backend. Best-effort
        — a logging failure must never affect the call, which is already over by the time we post."""
        if not self._intake_id:
            return  # a test call with no lead id — nothing to attach it to
        try:
            await self._post(
                "/agent/call-log",
                {"intakeId": self._intake_id, "transcript": transcript, "summary": summary},
            )
        except Exception as exc:  # noqa: BLE001
            log.warning("could not save call log: %s", exc)

    async def _post(self, path: str, body: dict) -> str:
        headers = {"Content-Type": "application/json"}
        if self._cfg.agent_tools_secret:
            headers["x-agent-secret"] = self._cfg.agent_tools_secret
        async with httpx.AsyncClient(timeout=self._cfg.request_timeout) as client:
            resp = await client.post(f"{self._cfg.backend_url}{path}", json=body, headers=headers)
        log.info("tool %s -> %s", path, resp.status_code)
        # Return the backend's JSON body verbatim; it carries a speakable `message` plus fields.
        return resp.text or json.dumps({"ok": resp.is_success})
